### PR TITLE
Add support for passing device into docker driver

### LINKED ISSUE #2938: Add support for "--device" option
We need an option that implements the "--device" Docker cli option, to mount usb devices and other serial control devices for hardware that container based applications are managing.

Per comment from @schmichael on #2671:

> Allowing mounting arbitrary devices makes sense. I think the distinction
between GPUs (#406) and other devices is we intend for the scheduler to
have first class knowledge of GPU resources. Other devices will just be
supported in driver configs.
>
> Jay or Pieter: Would one of you mind opening a Github issue for your
"--device" use cases? It shouldn't be blocked by #406.

### PR DESCRIPTION
**Pull request by preetapan**
Fixes #2938

### CODE REVIEW

File: client/driver/docker.go [hunk 1]
```diff
@@ -1020,6 +1046,20 @@ func (d *DockerDriver) createContainerConfig(ctx *ExecContext, task *structs.Tas
 		}
 	}
 
+	if len(driverConfig.Devices) > 0 {
+		var devices []docker.Device
+		for _, device := range driverConfig.Devices {
+			if device.HostPath != "" {
```
Comment (schmichael) [inline client/driver/docker.go]: Do we need to check this since validate will fail if HostPath is empty?

File: client/driver/docker_test.go [hunk 2]
```diff
@@ -1803,3 +1803,75 @@ func TestDockerDriver_OOMKilled(t *testing.T) {
...
+
+	ctx := testDockerDriverContexts(t, task)
+	driver := NewDockerDriver(ctx.DriverCtx)
+	copyImage(t, ctx.ExecCtx.TaskDir, "busybox.tar")
+	defer ctx.AllocDir.Destroy()
+
+	if _, err := driver.Prestart(ctx.ExecCtx, task); err != nil {
+		t.Fatalf("unexpected error:%v", err)
+	}
+
```
Comment (schmichael) [inline client/driver/docker_test.go]: Can we start and stat the file to make sure it exists? A lot of tests do this with the `waitForExist` helper and then inspecting the created container: https://github.com/hashicorp/nomad/blob/22400bffa2e36bc2d0c9a25ec151c2244c928321/client/driver/docker_test.go#L786

File: client/driver/docker.go [hunk 3]
```diff
@@ -190,13 +196,30 @@ type DockerDriverConfig struct {
...
+		for _, dev := range c.Devices {
+			if dev.HostPath == "" {
+				return fmt.Errorf("host path must be set in configuration for devices")
+			}
+
+			if dev.CgroupPermissions != "" {
+				for _, c := range dev.CgroupPermissions {
+					ch := string(c)
+					if ch != "r" && ch != "w" && ch != "m" {
+						return fmt.Errorf("invalid cgroup permission string: %q", dev.CgroupPermissions)
```
Comment (dadgar) [inline client/driver/docker.go]: Is this correct: https://github.com/docker/cli/blob/master/docs/reference/commandline/create.md#dealing-with-dynamically-created-devices---device-cgroup-rule?

Comment (preetapan) [reply client/driver/docker.go]: I made this validation logic after reading https://docs.docker.com/engine/reference/commandline/run/#add-host-device-to-container-device. Also, note this sentence at the end of that documentation: _Note: --device cannot be safely used with ephemeral devices. Block devices that may be removed should not be added to untrusted containers with --device._ The documentation you linked above for `--device-cgroup-rule` for dynamic devices, which looks like something else. I could be missing something though. 

We can create another follow up ticket to add configuration for `--device-cgroup-rule` without needing `--device` if this is something people need.

File: client/driver/docker_test.go (hunk 2 above)
Comment (preetapan) [reply client/driver/docker_test.go]: I validated that the `hostConfig` from docker inspect output showed the mounted device. Didn't do a stat of the device file. All the existing tests were reading things off the output of docker inspect so I did the same, this should verify that the device mounting did work.

File: client/driver/docker.go (hunk 3 above)
Comment (dadgar) [reply client/driver/docker.go]: Ah thanks for clarifying!

**Pull request merged**